### PR TITLE
Improve inference for string operations

### PR DESCRIPTION
**Pull request by timholy**
Together with https://github.com/JuliaIO/Tar.jl/pull/135, this eliminates nearly ~1400~>1700 invalidations mostly caused by InlineStrings.jl, and knocks about 300ms off the load time of CSV.jl (CC @quinnj).

More benefit still could be achieved from some version of https://github.com/JuliaLang/Pkg.jl/pull/3017 and #44486, but this already helps a lot.

### CODE REVIEW

File: base/logging.jl [hunk 1]
```diff
@@ -457,7 +457,7 @@ end
     msg = try
               "Exception while generating log record in module $_module at $filepath:$line"
           catch ex
-              "Exception handling log message: $ex"
+              "Exception handling log message: $(ex::Exception)"
```
Comment (vtjnash) [inline base/logging.jl:460]: exceptions are not necessary of type `Exception`

Comment (vtjnash) [reply base/logging.jl:460]: (We could potentially use LazyString, perhaps?)

File: base/process.jl [hunk 2]
```diff
@@ -69,7 +69,8 @@ function _uv_hook_close(proc::Process)
     nothing
 end
 
-const SpawnIOs = Vector{Any} # convenience name for readability
+const SpawnIO  = Redirectable
```
Comment (vtjnash) [inline base/process.jl:72]: This is a somewhat complicated union, which makes it non-ideal for asserting as the type. I think it also may be mostly deprecated, and mostly only present to help identify `FileRedirect` calls from other redirection calls.

File: base/show.jl [hunk 3]
```diff
@@ -1919,7 +1919,7 @@ function show_unquoted(io::IO, ex::Expr, indent::Int, prec::Int, quote_level::In
             na = length(func_args)
             if (na == 2 || (na > 2 && isa(func, Symbol) && func in (:+, :++, :*)) || (na == 3 && func === :(:))) &&
                     all(a -> !isa(a, Expr) || a.head !== :..., func_args)
-                sep = func === :(:) ? "$func" : " $func "
+                sep = func === :(:) ? "$func" : " " * repr(func) * " "   # if func::Any, avoid string interpolation (invalidation)
```
Comment (vtjnash) [inline base/show.jl:1922]: We have `func::Symbol` asserted just above

File: stdlib/REPL/src/LineEdit.jl [hunk 4]
```diff
@@ -1646,7 +1646,7 @@ end
 throw_eager_redirection_cycle(key::Union{Char, String}) =
     error("Eager redirection cycle detected for key ", repr(key))
 throw_could_not_find_redirected_value(key::Union{Char, String}) =
-    error("Could not find redirected value ", repl(key))
+    error("Could not find redirected value ", repr(key))
```
Comment (vtjnash) [inline stdlib/REPL/src/LineEdit.jl:1649]: :joy:

File: base/show.jl (hunk 3 above)
Comment (timholy) [reply base/show.jl:1922]: Only if `na > 2`?

Comment (vtjnash) [reply base/show.jl:1922]: Ah, right, the first argument call. `$` and `repr` do not produce the same result for Symbol though

Comment (timholy) [reply base/show.jl:1922]: Yeah, that was definitely a mistake.

**Comment by timholy** [comment]
I've been able to eliminate nearly all the invalidations (fewer than 100 left from an initial 1800) with this plus https://github.com/JuliaLang/Pkg.jl/pull/3020 and https://github.com/JuliaIO/Tar.jl/pull/136.

File: stdlib/REPL/src/TerminalMenus/AbstractMenu.jl [hunk 5]
```diff
@@ -192,7 +192,7 @@ function request(term::REPL.Terminals.TTYTerminal, m::AbstractMenu; cursor::Unio
         REPL.Terminals.raw!(term, true)
         true
     catch err
-        suppress_output || @warn("TerminalMenus: Unable to enter raw mode: $err")
+        suppress_output || @warn("TerminalMenus: Unable to enter raw mode: $(err::Exception)")
```
Comment (JeffBezanson) [inline stdlib/REPL/src/TerminalMenus/AbstractMenu.jl:195]: Here too, though I guess LazyString doesn't work here.

File: base/logging.jl [hunk 6]
```diff
@@ -674,7 +674,7 @@ function handle_message(logger::SimpleLogger, level::LogLevel, message, _module,
     end
     iob = IOContext(buf, stream)
     levelstr = level == Warn ? "Warning" : string(level)
-    msglines = eachsplit(chomp(string(message)::String), '\n')
+    msglines = eachsplit(chomp(convert(String, message)::String), '\n')
```
Comment (timholy) [inline base/logging.jl:677]: I'm the original source of the bug here. This and a similar change below in ConsoleLogger.jl should be backported to 1.8 (I'll submit a separate PR against that branch).

**Comment by timholy** [comment]
It appears (though there were a few other PRs too) that the combination of this PR and #44628 made `CairoMakie`'s load time drop from 8.2s to 7.1s. Nice outcome.

**Comment by timholy** [comment]
We may want to consider backporting this. See https://discourse.julialang.org/t/package-load-time-regressions-in-v1-8-beta3/78875?u=tim.holy (I don't know that the invalidation is the cause of the regression, but it's a bad sign).

**Comment by KristofferC** [comment]
Well, it would be worth testing if backporting this significantly improve things?

**Comment by timholy** [comment]
Sure, I guess if you're running PkgEval that pretty much gets rid of the risk. If it runs into trouble we could port the portion of this that is the "second half" of #43939, which if dim memories serve might fix the invalidation of `Base.require`.

**Pull request merged**